js/source/source.js

```diff
@@ -160,34 +114,28 @@ exports._querySourceFeatures = function(params) {
  * map.addSource('some id', sourceObj); // add
  * map.removeSource('some id');  // remove
  */
-exports.create = function(source) {
-    // This is not at file scope in order to avoid a circular require.
-    var sources = {
-        vector: require('./vector_tile_source'),
-        raster: require('./raster_tile_source'),
-        geojson: require('./geojson_source'),
-        video: require('./video_source'),
-        image: require('./image_source')
-    };
-
-    return exports.is(source) ? source : new sources[source.type](source);
+exports.create = function(id, source, dispatcher, onChange, callback) {
+    return exports.is(source) ? source :
+        sourceTypes[source.type](id, source, dispatcher, onChange, callback);
 };
 
 exports.is = function(source) {
-    // This is not at file scope in order to avoid a circular require.
-    var sources = {
-        vector: require('./vector_tile_source'),
-        raster: require('./raster_tile_source'),
-        geojson: require('./geojson_source'),
-        video: require('./video_source'),
-        image: require('./image_source')
-    };
+    // TODO: with the Source interface being factory-based, instanceof doesn't
+    // work as naturally for this.  Is Source.is() still needed?
+
+    return source && source.id;
+    // for (var type in sourceTypes) {
+    //     if (source instanceof sourceTypes[type]) {
+    //         return true;
+    //     }
+    // }
+    // return false;
+};
```

Comment:
We use `Source.is` to differentiate between these two cases:

```js
var source0 = new mapboxgl.GeoJSONSource(...);
map.addSource('foo', source0);

var source1 = { type: 'geojson' };
map.addSource('bar', source1);
```

I'm all for eliminating the method if possible. Exposing `GeoJSONSource` in our public API was a weird design choice but it'll be hard to go back on within the scope of this PR.